### ISSUE TITLE
Add alltoall collective communication support to torch.distributed module

### ISSUE THREAD

**Issue by ddkalamk** [pytorch/pytorch#32345]
## 🚀 Feature

I would request to add an Alltoall / alltoallv communication primitive to `torch.distributed` module and corresponding implementations to MPI, GLOO and other distributed backends.

## Motivation
When using both data and model parallelism in modern workloads (such as [DLRM](https://github.com/facebookresearch/dlrm/)), communication pattern required for switching from one type of parallelism to other maps very well to all-to-all communication well studied in HPC workloads. MPI provides [MPI_Alltoall](https://www.open-mpi.org/doc/v3.1/man3/MPI_Alltoall.3.php) (and more flexible [MPI_Alltoallv](https://www.open-mpi.org/doc/v3.1/man3/MPI_Alltoallv.3.php) APIs to perform this type of communication). In absence of alltoall support, we nee to use sequence of gathers or scatters to perform same communication. Having alltoall API allows reducing number of calls required to communication backend and avoid redundant data copies. 

[DLRM Research Paper](https://ai.facebook.com/blog/dlrm-an-advanced-open-source-deep-learning-recommendation-model/) explains this communication pattern as butterfly-shuffle in its _Parallelism_ section.

![image](https://user-images.githubusercontent.com/8791375/72598353-a73c0580-3935-11ea-84e3-03c61f4d4935.png)

## Pitch

Suppose we have 4 processes and each one has one `input` tensor with following data:

```
input:
[ 0  1  2  3]  # Proc 0
[ 4  5  6  7]  # Proc 1
[ 8  9 10 11]  # Proc 2
[12 13 14 15]  # Proc 3
```
After the communication, we want following output tensor on each processes:
```
output:
[ 0  4  8 12]  # Proc 0
[ 1  5  9 13]  # Proc 1
[ 2  6 10 14]  # Proc 2
[ 3  7 11 15]  # Proc 3
```

So, we simply call:

```
data = [ rank * 4 + i for i in range(4)]
input = torch.tensor(data)
output = torch.empty_like(input)
dist.alltoall(output, input)
```
This will split the input tensor equally and scatter to all the processes. On the other hand, output tensor would be split equally and receive gathered tensors from respective process.  

Alternatively, if we have list of tensors to scatter and gather list for receiving tensors or we explicitly want to specify what portion goes where (particularly, in multi-dimensional tensor), we can explicitly pass list of tensors as scatter_list and gather_list:
```
scatter_list = list(input.split(1))
gather_list = list(output.split(1))
dist.alltoall(gather_list, scatter_list)
```

## Alternatives
Today, doing it with scatter would require this:

```
scatter_list = list(input.split(1))
gather_list = list(output.split(1))
for i in range(4):
    dist.scatter(gather_list[i], scatter_list if i == rank else [], src = i)

```    

@mnaumovfb @JianpingChen066 @dmudiger @srinivas212 @Jianhui-Li @mshiryaev

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar

EDIT: Fixed order of input and output in alltoall call as per prototype code

**Comment by ddkalamk** [comment]
A prototype was done for required changes  this commit
https://github.com/ddkalamk/pytorch/commit/5b616536b10e063d13eefbe8b3de7eb96098af2c

**Comment by mrshenli** [comment]
Hey @ddkalamk, thanks for the pitch and implementation! I think this is a great addition to c10d. 
Would you like to create a PR to upstream https://github.com/ddkalamk/pytorch/commit/5b616536b10e063d13eefbe8b3de7eb96098af2c to PyTorch? We can continue discussions and reviews based on that code. 

BTW, how much perf gain did you see when using allToAll instead of multiple scatter-gather?

Loop in more people here for visibility. @pietern @zhaojuanmao @agolynski 

This will create conflict with #28068

**Comment by ddkalamk** [comment]
Hi @mrshenli, Thanks for the quick reply. I will submit a PR from my commit shortly.
Perf gain depends on backend used. However, a single call allows backend such as MPI to pick more efficient algorithm to perform communication.
I am not sure why these changes would conflict with #28068. I believe they are orthogonal modulo changes required in test to add new api.

BTW, just saw this #31892 . By same logic, we would need `ProcessGroup::alltoall` and `ProcessGroup::alltoall_base`. In fact, the implementation in https://github.com/ddkalamk/pytorch/commit/5b616536b10e063d13eefbe8b3de7eb96098af2c actually corresponds to `alltoall_base` (i.e. using single tensor instead of list of tensors)

**Comment by dmudiger** [comment]
Thanks @ddkalamk for putting this out and also the implementation.
 
We've been also testing the alltoall prototype implementation with DLRM at our end and have detailed perf analysis (@srinivas212 - has the details and can elaborate).

@dzhulgakov @manojkris - fyi from our previous conversation on this.

**Comment by mrshenli** [comment]
> BTW, just saw this #31892 . By same logic, we would need ProcessGroup::alltoall and ProcessGroup::alltoall_base. In fact, the implementation in ddkalamk@5b61653 actually corresponds to alltoall_base (i.e. using single tensor instead of list of tensors)

Yes, if we do the same thing as #31892, it wouldn't create conflict with #28068. (#28068 is defining what APIs a new backend should implement) But the current solution in #31892 in a temporary fix, and ultimately will move that API to `comm.cpp`. 

@agolynski @pietern  please correct me if I am wrong on the plan
